File: core/src/main/java/io/github/bindglam/neko/utils/Files.java
package io.github.bindglam.neko.utils;

import java.io.File;
import java.util.stream.Stream;

public final class Files {
    public static Stream<File> listFilesRecursively(File file) {
        try {
            return java.nio.file.Files.walk(file.toPath(), java.nio.file.FileVisitOption.FOLLOW_LINKS)
                    .filter(java.nio.file.Files::isRegularFile)
                    .map(java.nio.file.Path::toFile);
        } catch (java.io.IOException e) {
            return Stream.empty();
        }
    }

    private Files() {
    }
}

File: core/src/main/kotlin/com/bindglam/neko/utils/Files.kt
package com.bindglam.neko.utils

import com.bindglam.neko.pack.PackFile
import java.io.File
import java.nio.file.Files

fun File.createIfNotExists() {
    if(!parentFile.exists())
        parentFile.mkdirs()
    if(!exists())
        createNewFile()
}

fun File.listFilesRecursively(): List<File> = ArrayList<File>().apply {
    listFiles().forEach { file ->
        if(file.isDirectory)
            addAll(file.listFilesRecursively())
        else
            add(file)
    }
}

fun File.getRelativePath(root: String, newRoot: String, separator: String): String {
    val rootPath = if(newRoot.isEmpty()) "" else "${newRoot}${separator}"

    return "${rootPath}${if(path.startsWith(root)) path.substring(root.length+1).replace(File.separator, separator) else path.replace(File.separator, separator)}"
}

fun File.getRelativePath(root: String, separator: String): String = getRelativePath(root, "", separator)

fun File.getRelativePath(root: String): String = getRelativePath(root, File.separator)

fun File.toPackFile(): PackFile = PackFile({ inputStream().readBytes() }, Files.size(toPath()))